keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1204 @@
+     platform images, or to an SBOM artifact that refers to a platform matched
+     image. These artifacts may or may not be located on the same
+     registry/repository. The new referrers API allows for retrieving these
+     artifacts.
+
+**Validation:**
+   - Extend validation and security checks to cover new artifact types.
+
+**Storage Optimization in the container runtime:**
+   - Develop optimized storage solutions tailored for different artifact types,
+     potentially integrating with existing storage solutions or introducing new mechanisms.
+
+#### kubelet
+
+While the container runtime will be responsible of pulling and storing the OCI
+objects in the same way as for images, the kubelet still has to manage the full
+life-cycle of them. This means that some parts of the existing kubelet code can
+be reused, for example:
+
+- The logic how to ensure that an image exists on the node:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
+- The retrieval of available secrets for a pod:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988
+
+##### Pull Policy
+
+While the `imagePullPolicy` is working on container level, the introduced
+`pullPolicy` is a pod level construct. This means that we can support the same
+values `IfNotPresent`, `Always` and `Never`, but will only pull once per pod.
+
+Technically it means that we need to pull in [`SyncPod`](https://github.com/kubernetes/kubernetes/blob/b498eb9/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1049)
+for OCI objects on a pod level and not during [`EnsureImageExists`](https://github.com/kubernetes/kubernetes/blob/b498eb9/pkg/kubelet/images/image_manager.go#L102)
+before the container gets started.
+
+If users want to re-pull artifacts when referencing moving tags like `latest`,
+then they need to restart / evict the pod.
+
+##### Registry authentication
+
+For registry authentication purposes the same logic will be used as for the
+container image.
+
+#### CRI
+
+The CRI API is already capable of managing container images [via the `ImageService`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L146-L161).
+Those RPCs will be re-used for managing OCI artifacts, while the [`ImageSpec`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L798-L813)
+as well as [`PullImageResponse`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L1530-L1534)
+will be extended to mount the OCI object to a local path:
+
+```protobuf
+
+// ImageSpec is an internal representation of an image.
+message ImageSpec {
+    // …
+
+    // Indicate that the OCI object should be mounted.
+    bool mount = 20;
+
+    // SELinux label to be used.
+    string mount_label = 21;
+}
+
+message PullImageResponse {
+    // …
+
+    // Absolute local path where the OCI object got mounted.
+    string mountpoint = 2;
```

Comment:
mounted on a pull image request `with unpack` .. for a particular `pod image:count:tag (pid) + runtime handler`.. hmm

Reply:
if there are two pods with the same pull.. there could be two snapshots and could be two different versions if latest, or if the same sha could be optimized.. if two different runtime handlers there could be two snapshots.. 

Somehow we need to lock in the mount point and I'm not sure that is during pull image as that is not a part of the pod lifecycle.

This is a part of why on a container we may end up pulling at least a portion of the image again on create container even after the image has been ensured to be present by Kubelet. This was mentioned by Clayton in this kep discussion.

Reply:
I wish we could split runpodsandbox into create and start.. Until then my recommendation is pass in the pod image/artifact mount(s) in the runtpodsandbox response

Reply:
if there are two pods with the same pull.. there could be two snapshots and could be two different versions if latest, or if the same sha could be optimized.. if two different runtime handlers there could be two snapshots..

Somehow we need to lock in the mount point and I'm not sure that is during pull image as that is not a part of the pod lifecycle.

This is a part of why on a create container we check resolve, and if it can't resolve kubelet will have to ensure again, even after the image has been ensured a first time to be present by Kubelet. This was mentioned by Clayton in this kep discussion. At one point in history, we would pull the image a second time in create container vs ensure we can resolve and fail the create if resolve fails. https://github.com/containerd/containerd/blob/main/internal/cri/server/container_create.go#L107

I wish we could split runpodsandbox into create and start.. Until then my recommendation is return the pod image/artifact mount(s) in the runpodsandbox response

Reply:
Let's discuss that topic in real time (pinged you on slack including @haircommander and @mrunalp)

Reply:
The current design proposal still pulls on the `PullImage` RPC, but then passes the `Mount` as part of the `CreateContainer` call. This way we can fail if a mount does not exist and can ensure to have the same referring bind mounts on container creation.